### PR TITLE
[iOS] Unstake

### PR DESCRIPTION
**Pull request by thangpham26394**

### CODE REVIEW

File: EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift [hunk 1]
```diff
@@ -58,4 +118,89 @@ class UnstakeViewController: InAppBrowsingViewController {
...
+    func openTxStatusPopup(tx: PendingUnstakeTxInfo) {
+        let popup = StakingTrasactionProcessPopup.instantiateFromNib()
+        popup.tx = tx
+        let sheet = SheetViewController(controller: popup, sizes: [.fixed(420)], options: .init(pullBarHeight: 0))
+        dismiss(animated: true) {
+            UIApplication.shared.topMostViewController()?.present(sheet, animated: true)
+        }
+    }
+    
+    func approve() {
```
Comment (MrQKyber) [inline EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift:172]: Init and present view controllers inside view controller

File: EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift [hunk 2]
```diff
@@ -58,4 +118,89 @@ class UnstakeViewController: InAppBrowsingViewController {
...
+            showError()
+            return
+        } else {
+            hideError()
+        }
+        viewModel.unstakeValue = inputValue
+        receiveInfoView.setValue(value: viewModel.receivedValueString())
+    }
+    
+    func openUnStakeSummary() {
```
Comment (MrQKyber) [inline EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift:139]: Same here

File: EarnModule/EarnModule/Cells/StakingPortfolioCell.swift [hunk 3]
```diff
@@ -82,7 +79,7 @@ class StakingPortfolioCell: UITableViewCell {
   @IBOutlet weak var apyTitleLabel: UILabel!
   @IBOutlet weak var balanceTitleLabel: UILabel!
   
-  weak var delegate: StakingPortfolioCellDelegate?
+  var viewControler: UIViewController?
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Cells/StakingPortfolioCell.swift:82]: cell should not keep reference to UIViewController

Comment (tungnguyen20) [reply EarnModule/EarnModule/Cells/StakingPortfolioCell.swift:82]: Please use delegate or closure

File: EarnModule/EarnModule/Cells/StakingPortfolioCell.swift [hunk 4]
```diff
@@ -109,7 +106,7 @@ class StakingPortfolioCell: UITableViewCell {
   }
   
   @IBAction func inProcessButtonTapped(_ sender: UIButton) {
-    delegate?.warningButtonTapped()
+    viewControler?.showBottomBannerView(message: "It takes about x days to unstake. After that you can claim your rewards.")
```
[the remark this answers is not in the document]
Comment (tungnguyen20) [reply EarnModule/EarnModule/Cells/StakingPortfolioCell.swift:109]: x days ?? It should be specific days

File: EarnModule/EarnModule/Models/PendingStakingTxInfo.swift [hunk 5]
```diff
@@ -50,4 +50,7 @@ class PendingStakingTxInfo: PendingTxInfo {
         return ""
     }
 
+    override var amount: String? {
+        return platform.name.uppercased()
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Models/PendingStakingTxInfo.swift:54]: Look strange :-/

File: EarnModule/EarnModule/Models/PendingUnstakeTxInfo.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    override var description: String {
+        return "\(stakingTokenAmount) \(stakingTokenSymbol) → \(toTokenAmount) \(toTokenSymbol)"
+    }
+ 
+    override var detail: String {
+        return ""
+    }
+    
+    override var amount: String? {
+        return platform.name.uppercased()
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Models/PendingUnstakeTxInfo.swift:62]: Look strange

File: EarnModule/EarnModule/Resources/Strings.swift [hunk 7]
```diff
@@ -32,11 +34,22 @@ struct Strings {
...
     static let shouldNoMoreThan = "higher.than.max.stake.amount".toBeLocalised()
     static let shouldBeIntervalOf = "should.be.interval.of".toBeLocalised()
+    static let stakingInProgress = "staking.in.progress".toBeLocalised()
+    static let unstakeInProgress = "unstake.in.progress".toBeLocalised()
+    static let approveInProgress = "approve.in.progress".toBeLocalised()
+    static let success = "success".toBeLocalised()
+    static let txFailed = "tx.failed".toBeLocalised()
+    static let viewMyPool = "view.my.pool".toBeLocalised()
+    static let support = "support".toBeLocalised()
+    static let unstakeToken = "Unstake.token".toBeLocalised()
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Resources/Strings.swift:54]: Please lowercase keys

File: EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift [hunk 8]
```diff
@@ -22,29 +31,88 @@ class UnstakeViewController: InAppBrowsingViewController {
...
+    @IBAction func onBackButtonTapped(_ sender: Any) {
+        self.navigationController?.popViewController(animated: true)
+    }
+
     @IBAction func settingButtonTapped(_ sender: Any) {
-        
+        guard let viewModel = viewModel else { return }
+        TransactionSettingPopup.show(on: self, chain: viewModel.chain, currentSetting: viewModel.setting, onConfirmed: { [weak self] settingObject in
+            self?.viewModel?.setting = settingObject
+            self?.networkFeeView.setInfo(title: "Network Fee", value: viewModel.transactionFeeString())
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift:91]: Localize

File: EarnModule/EarnModule/Resources/Strings.swift [hunk 9]
```diff
@@ -57,6 +57,8 @@ struct Strings {
     static let searchPools = "search.pools".toBeLocalised()
     static let searchToken = "search.token".toBeLocalised()
     static let Staking = "Staking".toBeLocalised()
+    static let Stake = "Stake".toBeLocalised()
+    static let Unstake = "Unstake".toBeLocalised()
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/Resources/Strings.swift:61]: Please lowercase

File: EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    }
+    
+    func transactionFeeString() -> String {
+        return NumberFormatUtils.gasFee(value: setting.transactionFee(chain: chain)) + " " + AppState.shared.currentChain.quoteToken()
+    }
+    
+    func fetchData(controller: UIViewController) {
+        controller.showLoadingHUD()
+        apiService.getStakingOptionDetail(platform: platform.name, earningType: platform.type, chainID: "\(chain.getChainId())", tokenAddress: toUnderlyingTokenAddress) { result in
+            switch result {
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift:195]: Should call `fetchData(completion: @escaping () -> ())` and then handle show hide loading logic in ViewController

File: EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+            self.delegate?.didGetDataSuccess()
+        }
+    }
+    
+    func requestBuildUnstakeTx(showLoading: Bool = false, completion: @escaping ((Error?) -> Void) = {_ in }) {
+        apiService.buildUnstakeTx(param: buildTxRequestParams) { result in
+            switch result {
+            case .success(let tx):
+                self.txObject = tx
+                self.gasLimit = BigInt(tx.gasLimit.drop0x, radix: 16) ?? AppDependencies.gasConfig.earnGasLimitDefault
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift:237]: Set this gasLimit value to txObject  (refer to StakingViewController)

File: EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+                    self.delegate?.didGetDataSuccess()
+                }
+            case .failure(let error):
+                controller.hideLoading()
+                self.delegate?.didGetDataFail(errMsg: error.localizedDescription)
+            }
+        }
+    }
+    
+    func checkNeedApprove(earningToken: EarningToken, controller: UIViewController) {
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift:211]: Should not pass controller to this function as commented above

File: EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    }
+    
+    func openTxStatusPopup(tx: PendingUnstakeTxInfo, controller: UIViewController) {
+        let popup = StakingTrasactionProcessPopup.instantiateFromNib()
+        let viewModel = UnstakeTransactionProcessPopupViewModel(pendingStakingTx: tx)
+        popup.viewModel = viewModel
+        let sheet = SheetViewController(controller: popup, sizes: [.fixed(420)], options: .init(pullBarHeight: 0))
+        controller.navigationController?.popViewController(animated: true)
+        UIApplication.shared.topMostViewController()?.present(sheet, animated: true)
+    }
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift:168]: Open other ViewController should be call in ViewController

File: EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift [hunk 14]
```diff
@@ -22,29 +31,99 @@ class UnstakeViewController: InAppBrowsingViewController {
...
+        viewModel.unstakeValue = viewModel.balance
+        amountTextField.text = viewModel.unstakeValueString()
+        receiveInfoView.setValue(value: viewModel.receivedValueMaxString() + " " + viewModel.toTokenSymbol)
     }
     
     @IBAction func unstakeButtonTapped(_ sender: Any) {
-        showError()
+        switch unstakeButtonState {
+            case .normal:
+                viewModel?.openUnStakeSummary(controller: self)
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewControllers/UnstakeViewController.swift:118]: ViewController open itself

File: EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        if showRevertedRate {
+            let ratioString = NumberFormatUtils.balanceFormat(value: BigInt(10).power(36) / ratio, decimals: 18)
+            return "1 \(toTokenSymbol) = \(ratioString) \(stakingTokenSymbol)"
+        } else {
+            let ratioString = NumberFormatUtils.balanceFormat(value: ratio, decimals: 18)
+            return "1 \(stakingTokenSymbol) = \(ratioString) \(toTokenSymbol)"
+        }
+    }
+    
+    func approve(controller: UIViewController, onSuccess: @escaping (() -> Void), onFail: @escaping (() -> Void)) {
```
Comment (tungnguyen20) [inline EarnModule/EarnModule/ViewModel/UnstakeViewModel.swift:119]: Not pass controller here

**Review by tungnguyen20 — APPROVED** [review]

**Pull request merged**